### PR TITLE
Remove SupportedInnerType

### PR DESCRIPTION
**Pull request by coder0xff**
## Changes

Towards #1116 

This PR removes `SupportedInnerType` from src/parser/ty.rs. (Hopefully) in the future, `IrType::Unencodable` will be used to do things like using `RustOpaque` transparently, so "unsupported" types need to be able to propagate through the transformation stages. `SupportedInnerType` isn't of use in this situation, and seemed ripe for refactoring anyway, so it's been removed in favor of going from `syn::Type` directly to `IrType`. Type argument evaluation is now bottom-up. Additionally, sensitivity to fully qualified names has been added, but can be disabled by keeping the new "qualified_names" feature off.

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [x] The code generator is run and the code is formatted (via `just precommit`).
- [x] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [x] CI is passing.

## Remark for PR creator

- Justfile is a task runner for the command line interface that allows you to run shell commands from a file named `justfile` in your project directory. You can use Justfile after [installing it](https://github.com/casey/just). Note that commands written in `justfile` of this repository are expected to be run in `bash`, not `cmd` or `powershell`. Running `just ...` commands in `cmd` or `powershell` will produce errors as the syntax is not compatible. On Windows, you can use `git bash` if you have Git installed.
- If fzyzcjy does not reply for a few days, maybe he just did not see it, so please ping him.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
(given the title has "draft" I just marked it as draft. Feel free to mark it as ready for review later!)

**Comment by fzyzcjy** [comment]
Looks like `SupportedInnerType` is introduced in https://github.com/fzyzcjy/flutter_rust_bridge/pull/346. Thus cc @antonok-edm :)

**Review by fzyzcjy — COMMENTED** [review]
This looks like a nontrivial refactor, so I need to review it more carefully later (would be great to pass CI first). Good job!

File: frb_codegen/src/ir/ty_unencodable.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+//     lazy_static! {
+//         static ref OPAQUE_FILTER: Regex =
+//             Regex::new(r"(\bdyn|'static|\bDartSafe|\+ (Send|Sync|UnwindSafe|RefUnwindSafe))\b")
+//                 .unwrap();
+//     }
+//     OPAQUE_FILTER
+//         .replace_all(rust, "")
+//         .replace(char_not_alphanumeric, "_")
+//         .to_case(Case::Pascal)
+// }
```
Comment (antonok-edm) [inline frb_codegen/src/ir/ty_unencodable.rs:62]: cleanup?

File: frb_codegen/src/ir/ty_unencodable.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
+use crate::{ir::*, target::Target};
+use quote::ToTokens;
+
+// use convert_case::{Case, Casing};
+// use lazy_static::lazy_static;
+// use quote::ToTokens;
+// use regex::Regex;
```
Comment (antonok-edm) [inline frb_codegen/src/ir/ty_unencodable.rs:7]: cleanup?

**Review by antonok-edm — APPROVED** [review]
> Looks like `SupportedInnerType` is introduced in #346. Thus cc @antonok-edm :)

Yes, kill it with fire! :smiling_imp:

This is basically the direction I wanted to take the API in. #346 was my incremental "slightly better" improvement to what it was before. Thanks for taking it on, @coder0xff!

File: frb_codegen/src/ir/ty_unencodable.rs (hunk 1 above)
Comment (coder0xff) [reply frb_codegen/src/ir/ty_unencodable.rs]: fixed

File: frb_codegen/src/ir/ty_unencodable.rs (hunk 2 above)
Comment (coder0xff) [reply frb_codegen/src/ir/ty_unencodable.rs]: fixed

File: frb_codegen/Cargo.toml [hunk 3]
```diff
@@ -56,3 +56,5 @@ x86_64-pc-windows-msvc = {pkg-fmt = "zip"}
 chrono = []
 default = ["chrono"]
 uuid = []
+# When disabled, name qualifiers are discarded, e.g. my_library::SyncReturn == flutter_rust_bridge::SyncReturn
```
Comment (fzyzcjy) [inline frb_codegen/Cargo.toml:59]: uber-nit: curious why do we want that? IMHO nobody wants my_library::SyncReturn to be misunderstood as frb's SyncReturn

File: frb_codegen/src/ir/ty_unencodable.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+pub trait ArgsUnpackable {
+    fn unpack(&self) -> Vec<(&str, Option<ArgsRefs>)>;
+}
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct IrTypeUnencodable {
+    pub underlying_type: Box<syn::Type>,
+    pub segments: Vec<(String, Option<Args>)>,
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/ty_unencodable.rs:17]: uber-nit: shall we make the tuple a struct, then it is easier to read for readers

File: frb_codegen/src/ir/ty_unencodable.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub enum Args {
+    Generic(Vec<IrType>),
+    Signature(Vec<IrType>),
+}
+
+impl ArgsUnpackable for Vec<(String, Option<Args>)> {
+    fn unpack(&self) -> Vec<(&str, Option<ArgsRefs>)> {
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/ty_unencodable.rs:27]: uber-nit: looks like this wants to "take `&`"? then not sure why it is called "unpack"

**Review by fzyzcjy — COMMENTED** [review]
(partially reviewed, will continue soon, good job!)

File: frb_codegen/Cargo.toml (hunk 3 above)
Comment (fzyzcjy) [reply frb_codegen/Cargo.toml:59]: I see, the old behavior does so. Good new behavior!

File: frb_codegen/src/parser/ty.rs [hunk 6]
```diff
@@ -159,227 +104,368 @@ impl<'a> TypeParser<'a> {
...
-                        inner: Box::new(inner),
-                    })
-                }),
-                "Option" => {
-                    // Disallow nested Option
-                    if matches!(generic.as_ref(), SupportedInnerType:: Path(SupportedPathType { ident, .. }) if ident == "Option")
+
+                    [("DartOpaque", None)] => Ok(DartOpaque(IrTypeDartOpaque {})),
+
+                    [("String", None)] => Ok(Delegate(IrTypeDelegate::String)),
```
Comment (fzyzcjy) [inline frb_codegen/src/parser/ty.rs:279]: uber-nit: is it possible we avoid duplicating code for qualified_names? For example, is it possible we write down something like (super-pseudo code, hope you get the idea):

```rust
struct MyMatcher { crate_name, class_name }

impl MyMatcher {
  fn match(&self, flat_array) { flat_array == [crate_name, class_name] || flat_array == [class_name] }
}

let uuid_matcher = MyMatcher("uuid", "Uuid")

uuid_matcher.match(flat_array)
```

then no need to duplicate "uuid::Uuid" and "Uuid" as two separate cases

**Review by fzyzcjy — COMMENTED** [review]
Good job! Finished reviewing :)

File: frb_codegen/src/parser/ty.rs (hunk 6 above)
Comment (coder0xff) [reply frb_codegen/src/parser/ty.rs:279]: Would structural pattern matching still be possible? I think what we'd gain in DRYness we would lose tenfold in readability.

File: frb_codegen/src/ir/ty_unencodable.rs (hunk 5 above)
Comment (coder0xff) [reply frb_codegen/src/ir/ty_unencodable.rs]: Fixed, I called it "splay."

File: frb_codegen/src/parser/ty.rs (hunk 6 above)
Comment (fzyzcjy) [reply frb_codegen/src/parser/ty.rs:279]: Hmm looks reaasonable!

**Review by fzyzcjy — APPROVED** [review]
LGTM! Will merge after conflict is resolved and CI goes green

File: frb_codegen/src/parser/ty.rs (hunk 6 above)
Comment (coder0xff) [reply frb_codegen/src/parser/ty.rs:277]: For what it's worth, you could mark `not(feature = "qualified_names")` deprecated and clean them up for the next major version.

**Comment by fzyzcjy** [comment]
Nice refactor!

**Pull request merged**